MoreLinq.Test/BatchTest.cs

```diff
@@ -133,6 +133,16 @@ public void BatchEmptySource(SourceKind kind)
             var batches = Enumerable.Empty<int>().ToSourceKind(kind).Batch(100);
             Assert.That(batches, Is.Empty);
         }
+
+        [Test]
+        public void BatchUsesCollectionCountAtIterationTime()
+        {
+            var stack = new Stack<int>(Enumerable.Range(1, 3));
+            var result = stack.Batch(3);
+            stack.Push(4);
+            result.AssertCount(2).Consume();
+            Assert.Pass();
```

Comment:
This test doesn't look right. It doesn't exercise any of the optimisations for the various collection/list types. Those optimisations only kick in in two cases:

- The collection/list is empty.
- The collection/list count is smaller than the request batch size.

The test has a stack of 4 elements, which is greater than the request batch size of 3.

What you want here is something like the `BatchCollectionSmallerThanSize` test.

Reply:
Didn't realize that `Stack<>` doesn't implement `ICollection<T>`. Don't care about doing full test, just need to validate that the block of code is moved to the proper spot. Switched to `List<>`, which tests appropriately.

Reply:
> Don't care about doing full test, …

I don't know how to read that you “don't care”.

> just need to validate that the block of code is moved to the proper spot.

It's not about the physical movement of the code, but the observable behaviour that happens for a number of types at a certain time.

That aside, the test doesn't exercise the right path. It hits the `default` case in the `switch` within the iterator method. If you use a `BreakingList<>`, you'll see the test break:

```c#
[Test]
public void BatchUsesCollectionCountAtIterationTime()
{
    var list = Enumerable.Range(1, 3).ToList();
    var collection = new BreakingCollection<int>(list);
    var result = collection.Batch(3);
    list.Add(4);
    result.AssertCount(2).Consume();
    Assert.Pass();
}
```

Again:

> What you want here is something like the `BatchCollectionSmallerThanSize` test.

Reply:
Disagree. 

First, the test that you provided `BatchCollectionSmallerThanSizeUsesCollectionCountAtIterationTime` does not run correctly. All three classes (`BreakingCollection`, `BreakingList`, and `BreakingReadOnlyList`) copy their values internally at creation. Any changes to `list` are not propagated to these objects, so the test is done effectively on an unchanged list. `list.Add(3)` is a no-op.

Second, I have copied the code from the test here, to add additional conversation documentation. The important thing that are tracking is not the collection size itself. It is which branch of the `switch` are we using. Previously, the branching is done at the first call-site, so we receive an iterator that will always use `.CopyTo`, no matter whether the size has changed. The size of the array is queried at runtime, so that behavior is not in question here (i.e. if the collection size changes, it will use the appropriate collection size at runtime); however, if the size gets larger than the batch size, then the iterator _used_ should change to one that will split the collection correctly (aka the `default` branch). 

Under this PR, `Batch` correctly evaluates the `switch` at runtime, so when the size of the collection changes, the choice of iteration changes as well. _This_ is the behavior under test here. As such, it is not as important to test every branch under consideration; what's important is to test _that_ the branch selected changes.

Now, it may be important to also test that the collection count is obtained at runtime. In this case, we will need to add an option to `BreakingList` so that we can change the data inside of it after initialization. I'm happy to discuss the possibility of doing this. However, the test as written is correct, and tests the behavior that I intend to test; as such, it will remain.

```csharp
var list = new List<int>(Enumerable.Range(1, 3));

// verify that source <= batch size returns a single element
// since List<> implements ICollection<>, this should use
// `CopyTo` to generate the array.
var result = list.Batch(3);
result.AssertCount(1).Consume();

list.Add(4);

// verify that source > batch size returns two elements
// this test fails prior to PR #965;
// previously, `Batch` returned immediately the iterator
// method for `ICollection<> when Count <= size`, so
// it would return a single element with an array of size 4.
// corrected, `Batch` waits until iteration to decide which
// branch to use.
result.AssertCount(2).Consume();
Assert.Pass();
```

Reply:
Apologies - had more time to debug today.

The issue is that `.ToSourceKind` always does a `.ToList()`, which makes a copy. Your original test in this comment is valid, but the test as written in the code suggestion was not, due to the copy being made. 

Updated `.ToSourceKind` to accept a `bool copy`, which determines whether to make a copy or not. This way we can specify that we are not intending to copy, but to use the original `List<>` provided. This seemed the simplest way to update, though there are probably better ways.

Added your test, which correctly asserts that the collection count is obtained at runtime. Kept and modified my test to use `ToSourceKind` to validate that the branching is done at runtime.

Reply:
> The issue is that `.ToSourceKind` always does a `.ToList()`, which makes a copy. Your original test in this comment is valid, but the test as written in the code suggestion was not, due to the copy being made.

That's correct. I was in a hurry and only shared the changes to the test and figured you'd draw the logical conclusion about the changes needed to `ToSourceKind` from there. I could have made the guesswork easier by pointing it out.

Rather than add another flag to disable copying and do safe checks internally, why not go with distinct and strong-typed cases for each that read clearer:

```c#
        internal static IEnumerable<T> AsSourceKind<T>(this List<T> input, SourceKind sourceKind) =>
#pragma warning disable IDE0072 // Add missing cases
            sourceKind switch
#pragma warning restore IDE0072 // Add missing cases
            {
                SourceKind.BreakingList => new BreakingList<T>(input),
                SourceKind.BreakingReadOnlyList => new BreakingReadOnlyList<T>(input),
                SourceKind.BreakingCollection => new BreakingCollection<T>(input),
                SourceKind.BreakingReadOnlyCollection => new BreakingReadOnlyCollection<T>(input),
                _ => throw new ArgumentException(null, nameof(sourceKind))
            };

        internal static IEnumerable<T> ToSourceKind<T>(this IEnumerable<T> input, SourceKind sourceKind) =>
#pragma warning disable IDE0072 // Add missing cases
            sourceKind switch
#pragma warning restore IDE0072 // Add missing cases
            {
                SourceKind.Sequence => input.Select(x => x),
                var kind => input.ToList().AsSourceKind(kind)
            };
```

Here's the same as an overall patch that's ready for application (so there's no guesswork needed):

```diff
diff --git a/MoreLinq.Test/BatchTest.cs b/MoreLinq.Test/BatchTest.cs
index 51fb3360..4083cafb 100644
--- a/MoreLinq.Test/BatchTest.cs
+++ b/MoreLinq.Test/BatchTest.cs
@@ -140,7 +140,7 @@ public void BatchEmptySource(SourceKind kind)
         public void BatchCollectionSmallerThanSizeUsesCollectionCountAtIterationTime(SourceKind kind)
         {
             var list = new List<int> { 1, 2 };
-            var result = list.ToSourceKind(kind, copy: false).Batch(3);
+            var result = list.AsSourceKind(kind).Batch(3);
             list.Add(3);
 
             result.AssertSequenceEqual(new[] { 1, 2, 3 });
@@ -152,7 +152,7 @@ public void BatchCollectionSmallerThanSizeUsesCollectionCountAtIterationTime(Sou
         public void BatchUsesCollectionCountAtIterationTime(SourceKind kind)
         {
             var list = new List<int>(Enumerable.Range(1, 3));
-            var result = list.ToSourceKind(kind, copy: false).Batch(3);
+            var result = list.AsSourceKind(kind).Batch(3);
 
             // should use `CopyTo`
             result.AssertCount(1).Consume();
diff --git a/MoreLinq.Test/TestExtensions.cs b/MoreLinq.Test/TestExtensions.cs
index 95355de3..60f43889 100644
--- a/MoreLinq.Test/TestExtensions.cs
+++ b/MoreLinq.Test/TestExtensions.cs
@@ -87,18 +87,25 @@ internal static IEnumerable<string> GenerateSplits(this string str, params char[
                 yield return split;
         }
 
-        internal static IEnumerable<T> ToSourceKind<T>(this IEnumerable<T> input, SourceKind sourceKind, bool copy = true) =>
+        internal static IEnumerable<T> AsSourceKind<T>(this List<T> input, SourceKind sourceKind) =>
+#pragma warning disable IDE0072 // Add missing cases
             sourceKind switch
+#pragma warning restore IDE0072 // Add missing cases
             {
-                SourceKind.Sequence => copy ? input.Select(x => x) : throw new InvalidOperationException("Invalid to keep original as sequence."),
-                SourceKind.BreakingList => new BreakingList<T>(copy ? input.ToList() : input.SafeCast<List<T>>()),
-                SourceKind.BreakingReadOnlyList => new BreakingReadOnlyList<T>(copy ? input.ToList() : input.SafeCast<List<T>>()),
-                SourceKind.BreakingCollection => new BreakingCollection<T>(copy ? input.ToList() : input.SafeCast<List<T>>()),
-                SourceKind.BreakingReadOnlyCollection => new BreakingReadOnlyCollection<T>(copy ? input.ToList() : input.SafeCast<List<T>>()),
+                SourceKind.BreakingList => new BreakingList<T>(input),
+                SourceKind.BreakingReadOnlyList => new BreakingReadOnlyList<T>(input),
+                SourceKind.BreakingCollection => new BreakingCollection<T>(input),
+                SourceKind.BreakingReadOnlyCollection => new BreakingReadOnlyCollection<T>(input),
                 _ => throw new ArgumentException(null, nameof(sourceKind))
             };
 
-        internal static T SafeCast<T>(this object obj) where T : class =>
-            obj is T t ? t : throw new InvalidOperationException($"Must pass an object of type `{typeof(T)}`.");
+        internal static IEnumerable<T> ToSourceKind<T>(this IEnumerable<T> input, SourceKind sourceKind) =>
+#pragma warning disable IDE0072 // Add missing cases
+            sourceKind switch
+#pragma warning restore IDE0072 // Add missing cases
+            {
+                SourceKind.Sequence => input.Select(x => x),
+                var kind => input.ToList().AsSourceKind(kind)
+            };
     }
 }
```